src/common/wasm_util.cc

```diff
@@ -0,0 +1,159 @@
+
+#include "src/common/wasm_util.h"
+#include <cstdio>
+#include <string.h>
+
+namespace proxy_wasm {
+namespace common {
+
+bool WasmUtil::getCustomSection(const char *begin, const char *end, std::string_view name,
+                                std::string_view &ret) {
+  // Skip the Wasm header.
+  const char *pos = begin + 8;
+  while (pos < end) {
+    if (pos + 1 > end) {
+      return false;
+    }
+    const auto section_type = *pos++;
+    uint32_t section_len = 0;
+    if (!parseVarint(pos, end, section_len) || pos + section_len > end) {
+      return false;
+    }
+    if (section_type == 0) {
+      // Custom section.
+      const auto section_data_start = pos;
+      uint32_t section_name_len = 0;
+      if (!WasmUtil::parseVarint(pos, end, section_name_len) || pos + section_name_len > end) {
+        return false;
+      }
+      if (section_name_len == name.size() && ::memcmp(pos, name.data(), section_name_len) == 0) {
+        pos += section_name_len;
+        ret = {pos, static_cast<size_t>(section_data_start + section_len - pos)};
+        return true;
+      }
+      pos = section_data_start + section_len;
+    } else {
+      // Skip other sections.
+      pos += section_len;
+    }
+  }
+  return true;
+};
+
+bool WasmUtil::getFunctionNameIndex(const char *begin, const char *end,
+                                    std::unordered_map<uint32_t, std::string> &ret) {
+  std::string_view name_section = {};
+  if (!WasmUtil::getCustomSection(begin, end, "name", name_section)) {
+    return false;
+  };
+  if (!name_section.empty()) {
+    const char *pos = name_section.data();
+    const char *end = name_section.data() + name_section.size();
+    while (pos < end) {
+      const auto subsection_id = *pos++;
+      uint32_t subsection_size = 0;
+      if (!parseVarint(pos, end, subsection_size) || pos + subsection_size > end) {
+        return false;
+      }
+
+      if (subsection_id != 1) {
+        // Skip other subsctions.
+        pos += subsection_size;
+      } else {
+        // Enters function name subsection.
+        const auto start = pos;
+        uint32_t namemap_vector_size = 0;
+        if (!parseVarint(pos, end, namemap_vector_size) || pos + namemap_vector_size > end) {
+          return false;
+        }
+        for (uint32_t i = 0; i < namemap_vector_size; i++) {
+          uint32_t func_index = 0;
+          if (!parseVarint(pos, end, func_index)) {
+            return false;
+          }
+
+          uint32_t func_name_size = 0;
+          if (!parseVarint(pos, end, func_name_size) || pos + func_name_size > end) {
+            return false;
+          }
+          ret.insert({func_index, std::string(pos, func_name_size)});
+          pos += func_name_size;
+        }
+        if (start + subsection_size != pos) {
+          return false;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+bool WasmUtil::getStrippedSource(const char *begin, const char *end, std::vector<char> &ret) {
```

Comment:
Same here.

Reply:
Also, perhaps replace `std::vector<char>` with `std::vector<uint8_t>`, since it's more correct.